react/src/hooks/utils.tsx

```diff
@@ -1,14 +1,104 @@
 // Function patterns that commonly occur in this set of hooks.
 
+import { MutationKey, QueryClient } from "react-query";
+
+interface FetchOptions {
+    onSuccess?: (res: Response) => any;
+    onError?: (res: Response) => any;
+}
+
 /**
- * Fetch the provided url. Return the json response if successful.
+ * Fetch the provided url. Return the JSON response if successful.
  * Throw the response if unsuccessful.
  */
-export async function basicFetch(url: string) {
+export async function basicFetch(url: string, options?: FetchOptions) {
     const response = await fetch(url);
     if (response.ok) {
         return response.json();
     } else {
         throw response;
     }
 }
+
+/**
+ * Fetch the provided url with the given method and body (optional).
+ * Return the JSON response if successful.
+ * Throw the response if unsuccessful.
+ */
+export async function changeFetch(
+    url: string,
+    method: "POST" | "PATCH" | "DELETE",
+    body?: any,
+    overrides?: FetchOptions
+) {
+    const response = await fetch(url, {
+        method: method,
+        credentials: "same-origin",
+        headers: body ? { "Content-Type": "application/json" } : undefined,
+        body: body ? JSON.stringify(body) : undefined,
+    });
+    if (response.ok) {
+        if (overrides?.onSuccess) return overrides.onSuccess(response);
+        return response.json();
```

Comment:
You can actually use optional chaining with function calls too, but this might be confusing to read
```ts
overrides?.onSuccess?.(response) ?? await response.json();
```
So the return type of this handler is not necessarily the response because of the overrides, and it doesn't necessarily throw if onError doesn't throw?

Reply:
This is actually a good point regarding the typing. I'm going with the assumption for now that anyone who adds their own overrides will know what they're doing, but I can look into fixing the typing.